packages/aws-cdk/lib/api/cloudformation-deployments.ts

```diff
@@ -179,14 +179,31 @@ export interface ProvisionerProps {
  */
 export class CloudFormationDeployments {
   private readonly sdkProvider: SdkProvider;
+  private readonly ssmCache = new Map<string, number>();
 
   constructor(props: ProvisionerProps) {
     this.sdkProvider = props.sdkProvider;
   }
 
   public async readCurrentTemplate(stackArtifact: cxapi.CloudFormationStackArtifact): Promise<Template> {
     debug(`Reading existing template for stack ${stackArtifact.displayName}.`);
-    const { stackSdk } = await this.prepareSdkFor(stackArtifact, undefined, Mode.ForReading);
+    let stackSdk: ISDK | undefined = undefined;
+    // try to assume the lookup role and fallback to the deploy role
+    try {
+      const result = await this.prepareSdkWithLookupRoleFor(stackArtifact);
+      if (!result.defaultCredentials) {
+        stackSdk = result.stackSdk;
+      } else {
+        warning(`${stackArtifact.lookupRole?.arn} could not be assumed. Please upgrade to bootstrap stack >= ${stackArtifact.lookupRole?.requiresBootstrapStackVersion}`);
+      }
+    } catch (e) {
+      warning(`${stackArtifact.lookupRole?.arn} could not be assumed. Please upgrade to bootstrap stack >= ${stackArtifact.lookupRole?.requiresBootstrapStackVersion}`);
```

Comment:
Should we at least log the exception we caught? At `debug` level or something?

Reply:
I also think this message should probably be moved into `prepareSdkWithLookupRoleFor` -- otherwise every location that wants to call `prepareSdkWithLookupRoleFor` should add this check.

Reply:
And let's have the message inform users that we're proceeding with normal credentials, that if everything works they don't need to do anything, and the upgrade is only necessary to get rid of the warning.

Reply:
So for the messaging we have a couple scenarios:

1. assuming lookup role failed, default credentials are returned
In this case `forEnvironment` will log 2 messages. A debug with the error message and a warning with something like:
```
baseCreds could not be used to assume lookupRoleArn, but are for the right account. Proceeding anyway.
```
In `prepareSdkWithLookupRoleFor` we could add another warning with:
```
To get rid of this warning, please upgrade to bootstrap version >= 8.
```

2. assuming lookup role succeeded, but bootstrap stack version is not valid &
3. assuming lookup role failed _and_ default credentials are not correct (i.e. `forEnvironment` throws an error)

In `prepareSdkWithLookupRoleFor` we could add debug with the error message and  a warning with:
```
Could not assume lookupRoleArn. Proceeding anyway.
To get rid of this warning, please upgrade to bootstrap version >= 8.
```

What do you think?

Reply:
Sounds good if I'm understanding what you're saying!

Reply:
Put it between parens?

```
Could not assume lookupRoleArn. Proceeding anyway.
(To get rid of this warning, please upgrade to bootstrap version >= 8.)
```

Reply:
Updated.